File: killrvideo-service-comments/src/test/java/com/killrvideo/service/comment/grpc/CommentsServiceGrpcTest.java
package com.killrvideo.service.comment.grpc;

import com.killrvideo.dse.dto.ResultListPage;
import com.killrvideo.messaging.dao.MessagingDao;
import com.killrvideo.service.comment.dto.Comment;
import com.killrvideo.service.comment.dto.QueryCommentByUser;
import com.killrvideo.service.comment.dto.QueryCommentByVideo;
import com.killrvideo.service.comment.repository.CommentRepository;
import io.grpc.stub.StreamObserver;
import killrvideo.comments.CommentsServiceOuterClass.*;
import killrvideo.comments.events.CommentsEvents;
import org.junit.jupiter.api.*;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import java.util.concurrent.CompletableFuture;

import static org.mockito.Mockito.*;

@SuppressWarnings("unchecked")
public class CommentsServiceGrpcTest {
    @InjectMocks
    private CommentsServiceGrpc service;
    @Mock
    private CommentRepository commentRepository;
    @Mock
    private MessagingDao messagingDao;
    @Mock
    private CommentsServiceGrpcValidator validator;
    @Mock
    private CommentsServiceGrpcMapper mapper;

    private AutoCloseable closeable;

    @BeforeEach
    public void openMocks() {
        closeable = MockitoAnnotations.openMocks(this);
    }

    @AfterEach
    public void releaseMocks() throws Exception {
        closeable.close();
    }

    @Test
    public void testCommentOnVideoWithValidationFailed() {
        CommentOnVideoRequest grpcReq = CommentOnVideoRequest.getDefaultInstance();
        StreamObserver<CommentOnVideoResponse> grpcResObserver = mock(StreamObserver.class);

        doThrow(new IllegalArgumentException()).when(validator).validateGrpcRequestCommentOnVideo(any(),any());

        Assertions.assertThrows(IllegalArgumentException.class, () ->
                service.commentOnVideo(grpcReq, grpcResObserver)
        );
    }

    @Test
    public void testCommentOnVideoAllWithInsertFailed() {
        CommentOnVideoRequest grpcReq = CommentOnVideoRequest.getDefaultInstance();
        StreamObserver<CommentOnVideoResponse> grpcResObserver = mock(StreamObserver.class);

        doNothing().when(validator).validateGrpcRequestCommentOnVideo(any(),any());
        Comment comment = mock(Comment.class);
        CommentsEvents.UserCommentedOnVideo event = CommentsEvents.UserCommentedOnVideo.getDefaultInstance();

        when(mapper.mapToComment(any())).thenReturn(comment);
        when(mapper.createUserCommentedOnVideoEvent(any())).thenReturn(event);
        when(commentRepository.insertCommentAsync(comment))
                .thenReturn(CompletableFuture.failedFuture(new Exception()));

        service.commentOnVideo(grpcReq, grpcResObserver);
        verify(messagingDao, times(0)).sendEvent(any(), any());
        verify(grpcResObserver, times(1)).onError(any());
        verify(grpcResObserver, times(0)).onNext(any());
        verify(grpcResObserver, times(0)).onCompleted();
    }

    @Test
    public void testCommentOnVideoAllWithSendFailed() {
        CommentOnVideoRequest grpcReq = CommentOnVideoRequest.getDefaultInstance();
        StreamObserver<CommentOnVideoResponse> grpcResObserver = mock(StreamObserver.class);

        doNothing().when(validator).validateGrpcRequestCommentOnVideo(any(),any());
        Comment comment = mock(Comment.class);
        CommentsEvents.UserCommentedOnVideo event = CommentsEvents.UserCommentedOnVideo.getDefaultInstance();

        when(mapper.mapToComment(any())).thenReturn(comment);
        when(mapper.createUserCommentedOnVideoEvent(any())).thenReturn(event);
        when(commentRepository.insertCommentAsync(comment))
                .thenReturn(CompletableFuture.completedFuture(comment));
        when(messagingDao.sendEvent(any(), any())).thenReturn(CompletableFuture.failedFuture(new Exception()));

        service.commentOnVideo(grpcReq, grpcResObserver);
        verify(grpcResObserver, times(1)).onError(any());
        verify(grpcResObserver, times(0)).onNext(any());
        verify(grpcResObserver, times(0)).onCompleted();
    }

    @Test
    public void testCommentOnVideo() {
        CommentOnVideoRequest grpcReq = CommentOnVideoRequest.getDefaultInstance();
        StreamObserver<CommentOnVideoResponse> grpcResObserver = mock(StreamObserver.class);

        doNothing().when(validator).validateGrpcRequestCommentOnVideo(any(),any());
        Comment comment = mock(Comment.class);
        CommentsEvents.UserCommentedOnVideo event = CommentsEvents.UserCommentedOnVideo.getDefaultInstance();

        when(mapper.mapToComment(any())).thenReturn(comment);
        when(mapper.createUserCommentedOnVideoEvent(any())).thenReturn(event);
        when(commentRepository.insertCommentAsync(comment))
                .thenReturn(CompletableFuture.completedFuture(comment));
        when(messagingDao.sendEvent(any(), any())).thenReturn(CompletableFuture.completedFuture(null));

        service.commentOnVideo(grpcReq, grpcResObserver);
        verify(grpcResObserver, times(0)).onError(any());
        verify(grpcResObserver, times(1)).onNext(any());
        verify(grpcResObserver, times(1)).onCompleted();
    }

    @Test
    public void testGetUserCommentsWithValidationFailed() {
        GetUserCommentsRequest grpcReq = GetUserCommentsRequest.getDefaultInstance();
        StreamObserver<GetUserCommentsResponse> responseObserver = mock(StreamObserver.class);

        doThrow(new IllegalArgumentException()).when(validator).validateGrpcRequest_GetUserComments(any(),any());

        Assertions.assertThrows(IllegalArgumentException.class, () ->
                service.getUserComments(grpcReq, responseObserver)
        );
    }

    @Test
    public void testGetUserCommentsWithQueryFailed() {
        GetUserCommentsRequest grpcReq = GetUserCommentsRequest.getDefaultInstance();
        StreamObserver<GetUserCommentsResponse> responseObserver = mock(StreamObserver.class);

        doNothing().when(validator).validateGrpcRequest_GetUserComments(any(),any());
        QueryCommentByUser query = mock(QueryCommentByUser.class);
        when(mapper.mapFromGrpcUserCommentToDseQuery(any())).thenReturn(query);
        when(commentRepository.findCommentsByUserIdAsync(any())).thenReturn(CompletableFuture.failedFuture(new Exception()));

        this.service.getUserComments(grpcReq, responseObserver);
        verify(responseObserver, times(1)).onError(any());
        verify(responseObserver, times(0)).onNext(any());
        verify(responseObserver, times(0)).onCompleted();
    }

    @Test
    public void testGetUserComments() {
        GetUserCommentsRequest grpcReq = GetUserCommentsRequest.getDefaultInstance();
        StreamObserver<GetUserCommentsResponse> responseObserver = mock(StreamObserver.class);

        doNothing().when(validator).validateGrpcRequest_GetUserComments(any(),any());
        QueryCommentByUser query = mock(QueryCommentByUser.class);
        when(mapper.mapFromGrpcUserCommentToDseQuery(any())).thenReturn(query);

        ResultListPage<Comment> resultListPage = mock(ResultListPage.class);
        when(commentRepository.findCommentsByUserIdAsync(any())).thenReturn(
                CompletableFuture.completedFuture(resultListPage)
        );

        this.service.getUserComments(grpcReq, responseObserver);
        verify(responseObserver, times(0)).onError(any());
        verify(responseObserver, times(1)).onNext(any());
        verify(responseObserver, times(1)).onCompleted();
    }
    @Test
    public void testGetVideoCommentsWithValidationFailed() {
        GetVideoCommentsRequest grpcReq = GetVideoCommentsRequest.getDefaultInstance();
        StreamObserver<GetVideoCommentsResponse> responseObserver = mock(StreamObserver.class);

        doThrow(new IllegalArgumentException()).when(validator).validateGrpcRequestGetVideoComment(any(),any());

        Assertions.assertThrows(IllegalArgumentException.class, () ->
                service.getVideoComments(grpcReq, responseObserver)
        );
    }

    @Test
    public void testGetVideoCommentsWithQueryFailed() {
        GetVideoCommentsRequest grpcReq = GetVideoCommentsRequest.getDefaultInstance();
        StreamObserver<GetVideoCommentsResponse> responseObserver = mock(StreamObserver.class);

        doNothing().when(validator).validateGrpcRequestGetVideoComment(any(),any());
        QueryCommentByVideo query = mock(QueryCommentByVideo.class);
        when(mapper.mapFromGrpcVideoCommentToDseQuery(any())).thenReturn(query);
        when(commentRepository.findCommentsByVideosIdAsync(any())).thenReturn(CompletableFuture.failedFuture(new Exception()));

        this.service.getVideoComments(grpcReq, responseObserver);
        verify(responseObserver, times(1)).onError(any());
        verify(responseObserver, times(0)).onNext(any());
        verify(responseObserver, times(0)).onCompleted();
    }

    @Test
    public void testGetVideoComments() {
        GetVideoCommentsRequest grpcReq = GetVideoCommentsRequest.getDefaultInstance();
        StreamObserver<GetVideoCommentsResponse> responseObserver = mock(StreamObserver.class);

        doNothing().when(validator).validateGrpcRequestGetVideoComment(any(),any());
        QueryCommentByVideo query = mock(QueryCommentByVideo.class);
        when(mapper.mapFromGrpcVideoCommentToDseQuery(any())).thenReturn(query);

        ResultListPage<Comment> resultListPage = mock(ResultListPage.class);
        when(commentRepository.findCommentsByVideosIdAsync(any())).thenReturn(CompletableFuture.completedFuture(resultListPage));

        this.service.getVideoComments(grpcReq, responseObserver);
        verify(responseObserver, times(0)).onError(any());
        verify(responseObserver, times(1)).onNext(any());
        verify(responseObserver, times(1)).onCompleted();
    }
}

File: killrvideo-service-comments/src/test/kotlin/com/killrvideo/service/comment/grpc/CommentsServiceGrpcTest.kt
package com.killrvideo.service.comment.grpc

import com.killrvideo.dse.dto.ResultListPage
import com.killrvideo.messaging.dao.MessagingDao
import com.killrvideo.service.comment.dto.Comment
import com.killrvideo.service.comment.repository.CommentRepository
import com.killrvideo.utils.GrpcMappingUtils.randomUuid
import io.mockk.*
import io.mockk.impl.annotations.InjectMockKs
import io.mockk.impl.annotations.MockK
import killrvideo.comments.CommentsServiceOuterClass.GetUserCommentsResponse
import killrvideo.comments.CommentsServiceOuterClass.GetVideoCommentsResponse
import killrvideo.comments.commentOnVideoRequest
import killrvideo.comments.events.userCommentedOnVideo
import killrvideo.comments.getUserCommentsRequest
import killrvideo.comments.getVideoCommentsRequest
import kotlinx.coroutines.runBlocking
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import java.util.concurrent.CompletableFuture

class CommentsServiceGrpcTest {
    @InjectMockKs
    private lateinit var service: CommentsServiceGrpc

    @MockK
    private lateinit var commentRepository: CommentRepository

    @MockK
    private lateinit var messagingDao: MessagingDao

    @MockK
    private lateinit var validator: CommentsServiceGrpcValidator

    @MockK
    private lateinit var mapper: CommentsServiceGrpcMapper
    private val serviceKey = "CommentsService"
    private val messageDestination = "topic-kv-commentCreation"

    @BeforeEach
    fun setUp() = MockKAnnotations.init(this, relaxUnitFun = true)

    @Test
    fun testCommentOnVideoWithValidationFailed() {
        val request = commentOnVideoRequest {}

        every {
            validator.validateGrpcRequestCommentOnVideo(any())
        } throws IllegalArgumentException()

        assertThrows<IllegalArgumentException> {
            runBlocking {
                service.commentOnVideo(request)
            }
        }
    }

    @Test
    fun testCommentOnVideoAllWithInsertFailed() {
        val request = commentOnVideoRequest {}
        every {
            validator.validateGrpcRequestCommentOnVideo(any())
        } just Runs

        val event = userCommentedOnVideo {}
        every { mapper.createUserCommentedOnVideoEvent(any())} returns event
        coEvery { commentRepository.insertCommentAsync(any()) } throws Exception()

        assertThrows<Exception> {
            runBlocking { service.commentOnVideo(request) }
        }

        verify(exactly = 0) { messagingDao.sendEvent(any(), any()) }
    }

    @Test
    fun testCommentOnVideoAllWithSendFailed() {
        val request = commentOnVideoRequest {}
        every {
            validator.validateGrpcRequestCommentOnVideo(any())
        } just Runs
        val comment = mockk<Comment>()
        val event = userCommentedOnVideo {}

        every {mapper.createUserCommentedOnVideoEvent(any())} returns event
        coEvery { commentRepository.insertCommentAsync(any()) } returns comment
        val future: CompletableFuture<Void> = CompletableFuture.failedFuture(Exception())
        every {messagingDao.sendEvent(any(), any()) } returns future

        assertThrows<Exception> {
            runBlocking { service.commentOnVideo(request) }
        }
    }

    @Test
    fun testCommentOnVideo() {
        val request = commentOnVideoRequest {}
        every {
            validator.validateGrpcRequestCommentOnVideo(any())
        } just Runs
        val comment = mockk<Comment>()
        val event = userCommentedOnVideo {}

        every {mapper.createUserCommentedOnVideoEvent(any())} returns event
        coEvery { commentRepository.insertCommentAsync(any()) } returns comment
        every {messagingDao.sendEvent(any(), any()) } returns
                CompletableFuture.completedFuture(null)
        runBlocking { service.commentOnVideo(request) }
    }

    @Test
    fun testGetUserCommentsWithValidationFailed() {
        val request = getUserCommentsRequest {}
        every {
            validator.validateGrpcRequestCommentOnVideo(any())
        } throws IllegalArgumentException()
        assertThrows<IllegalArgumentException> {
            runBlocking { service.getUserComments(request) }
        }
    }

    @Test
    fun testGetUserCommentsWithQueryFailed() {
        val request = getUserCommentsRequest {}
        every {
            validator.validateGrpcRequestCommentOnVideo(any())
        } just Runs

        coEvery {
            commentRepository.findCommentsByUserIdAsync(any())
        } throws Exception()
        assertThrows<Exception> {
            runBlocking { service.getUserComments(request) }
        }
    }

    @Test
    fun testGetUserComments() {
        val request = getUserCommentsRequest {
            userId = randomUuid()
            pageSize = 2
        }
        every {
            validator.validateGrpcRequestCommentOnVideo(any())
        } just Runs
        val resultListPage: ResultListPage<Comment> = mockk()
        coEvery {
            commentRepository.findCommentsByUserIdAsync(any())
        } returns resultListPage

        val response = mockk<GetUserCommentsResponse>()
        every { mapper.mapFromDseUserCommentToGrpcResponse(any()) } returns response

        val result = runBlocking { service.getUserComments(request) }
        assertEquals(response, result)
    }

    @Test
    fun testGetVideoCommentsWithValidationFailed() {
        val request = getVideoCommentsRequest {}
        every {
            validator.validateGrpcRequestCommentOnVideo(any())
        } throws IllegalArgumentException()
        assertThrows<IllegalArgumentException> {
            runBlocking { service.getVideoComments(request) }
        }
    }

    @Test
    fun testGetVideoCommentsWithQueryFailed() {
        val request = getVideoCommentsRequest {}
        every { validator.validateGrpcRequestGetVideoComment(any()) } just Runs
        coEvery {
            commentRepository.findCommentsByVideosIdAsync(any())
        } throws Exception()

        assertThrows<Exception> {
            runBlocking {
                service.getVideoComments(request)
            }
        }
    }

    @Test
    fun testGetVideoComments() {
        val request = getVideoCommentsRequest {
            videoId = randomUuid()
            pageSize = 2
        }

        every { validator.validateGrpcRequestGetVideoComment(any()) } just Runs

        val resultListPage: ResultListPage<Comment> = mockk()
        coEvery {
            commentRepository.findCommentsByVideosIdAsync(any())
        } returns resultListPage
        val response = mockk<GetVideoCommentsResponse>()
        every { mapper.mapFromDseVideoCommentToGrpcResponse(any()) } returns response
        val result = runBlocking {
            service.getVideoComments(request)
        }

        assertEquals(response, result)
    }
}
